src/Microsoft.AspNetCore.Blazor.Razor.Extensions/BlazorRuntimeNodeWriter.cs

```diff
@@ -373,6 +363,173 @@ private void WriteElementAttributeToken(CodeRenderingContext context, HtmlTagTok
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (node == null)
+            {
+                throw new ArgumentNullException(nameof(node));
+            }
+
+            _scopeStack.OpenScope(node.TagName, isComponent: true);
+            context.RenderChildren(node);
+            _scopeStack.CloseScope(context, node.TagName, isComponent: true, source: node.Source);
+        }
+
+        public override void WriteComponentAttribute(CodeRenderingContext context, ComponentAttributeExtensionNode node)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (node == null)
+            {
+                throw new ArgumentNullException(nameof(node));
+            }
+
+            // builder.OpenComponent<TComponent>(42);
+            context.CodeWriter.Write(_scopeStack.BuilderVarName);
+            context.CodeWriter.Write(".");
+            context.CodeWriter.Write(BlazorApi.RenderTreeBuilder.AddAttribute);
+            context.CodeWriter.Write("(");
+            context.CodeWriter.Write((_sourceSequence++).ToString());
+            context.CodeWriter.Write(", ");
+            context.CodeWriter.WriteStringLiteral(node.AttributeName);
+            context.CodeWriter.Write(", ");
+
+            if (node.AttributeStructure == AttributeStructure.Minimized)
+            {
+                // Minimized attributes always map to 'true'
+                context.CodeWriter.Write("true");
+            }
+            else if (
+                node.Children.Count != 1 ||
+                node.Children[0] is HtmlContentIntermediateNode htmlNode && htmlNode.Children.Count != 1 ||
+                node.Children[0] is CSharpExpressionIntermediateNode cSharpNode && cSharpNode.Children.Count != 1)
+            {
+                // We don't expect this to happen, we just want to know if it can.
+                throw new InvalidOperationException("Attribute nodes should either be minimized or a single content node.");
+            }
+            else if (node.BoundAttribute.IsUIEventHandlerProperty())
+            {
+                // This is a UIEventHandler property. We do some special code generation for this
+                // case so that it's easier to write for common cases.
+                //
+                // Example: 
+                //      <MyComponent OnClick="Foo()"/> 
+                //      --> builder.AddAttribute(X, "OnClick", new UIEventHandler((e) => Foo()));
+                //
+                // The constructor is important because we want to put type inference into a state where
+                // we know the delegate's type should be UIEventHandler. AddAttribute has an overload that
+                // accepts object, so without the 'new UIEventHandler' things will get ugly.
```

Comment:
I know this is a general C# question but after compilation, is `new UIEventHandler((e) => Foo())` equivalent to `((UIEventHandler)((e) => Foo()))`? The extra `new` doesn't cause an extra heap allocation, does it?

Reply:
I believe they both allocate, but I'm not totally sure.

I know the former does the right thing for type inference on the lambda, which is important for design time but not runtime. I'll look into this when I revisit the delegate to make it better. We can make the runtime code as efficient as possible while keeping design time tailored to the intellisense experience.